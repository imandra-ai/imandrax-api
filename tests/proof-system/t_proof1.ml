module P = Imandra_proof_system_encode
module E_cbor = Imandra_proof_system_encode_cbor

let () =
  let buf = Buffer.create 32 in
  let enc = E_cbor.encoder in

  let total_bytes = ref 0 in
  Printf.printf "start proof\n";

  (* make an output that shows an hexdump per entry *)
  let output_entry buf =
    let str = Buffer.contents buf in
    total_bytes := !total_bytes + String.length str;
    let cbor, _ = CBOR.Simple.decode_partial str in
    Printf.printf "entry (%dB): %s\n" (String.length str)
      (CBOR.Simple.to_diagnostic cbor);
    (* (CBOR.SiHex.hexdump_s @@ Hex.of_string @@ Buffer.contents buf); *)
    Buffer.clear buf
  in
  let out =
    P.Output.(Out { st = buf; enc; output_entry; gen_id = make_gen_id () })
  in

  let id_bool = P.ty_app out "bool" [] in
  let id_true = P.t_bool out true in

  let id_x = P.t_var out (P.var out "x" id_bool) in
  let id_x_eq_x = P.t_builtin_eq out id_bool id_x id_x in
  let _id_ass = P.p_assume out P.scope0 id_x_eq_x in

  Printf.printf "proof done (%d B total)\n%!" !total_bytes;
  ()
