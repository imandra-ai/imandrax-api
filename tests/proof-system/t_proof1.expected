start proof
entry (13B): ["ty.@", 0, "bool", []]
entry (10B): ["t.bool", 1, true]
entry (9B): ["var", 2, "x", 0]
entry (7B): ["t.v", 3, 2]
entry (13B): ["t.b", 4, "=", [0], [3, 3]]
entry (13B): ["p.assume", 5, 0, 4]
proof done (65 B total)
